File: Sources/FWObjC/Module/View/DrawerView.m
//
//  DrawerView.m
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

#import "DrawerView.h"
#import <objc/runtime.h>

#if FWMacroSPM

@interface UIView ()

@property (nonatomic, strong, nullable) __FWDrawerView *__fw_drawerView;

@end

@interface UIScrollView ()

@property (nonatomic, assign, readonly) BOOL __fw_canScrollVertical;
@property (nonatomic, assign, readonly) BOOL __fw_canScrollHorizontal;
- (BOOL)__fw_isScrollTo:(UIRectEdge)edge;
- (void)__fw_scrollTo:(UIRectEdge)edge animated:(BOOL)animated;

@end

#else

#import <FWFramework/FWFramework-Swift.h>

#endif

#pragma mark - __FWDrawerView

@interface __FWDrawerView () <UIGestureRecognizerDelegate, UIScrollViewDelegate>

@property (nonatomic, assign) CGFloat position;
@property (nonatomic, strong) CADisplayLink *displayLink;
@property (nonatomic, assign) BOOL panDisabled;
@property (nonatomic, assign) CGFloat originPosition;
@property (nonatomic, assign) CGPoint originOffset;
@property (nonatomic, assign) BOOL isOriginDraggable;
@property (nonatomic, assign) BOOL isOriginScrollable;
@property (nonatomic, assign) BOOL isOriginScrollView;
@property (nonatomic, assign) BOOL isOriginDirection;

@end

@implementation __FWDrawerView

#pragma mark - Lifecycle

- (instancetype)initWithView:(UIView *)view
{
    self = [super init];
    if (self) {
        _view = view;
        _autoDetected = YES;
        _kickbackHeight = 0;
        _positions = @[];
        _direction = UISwipeGestureRecognizerDirectionUp;
        _position = self.isVertical ? view.frame.origin.y : view.frame.origin.x;
        if ([view isKindOfClass:[UIScrollView class]]) {
            self.scrollView = (UIScrollView *)view;
        }
        
        UIPanGestureRecognizer *gestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(gestureRecognizerAction:)];
        _gestureRecognizer = gestureRecognizer;
        gestureRecognizer.delegate = self;
        [view addGestureRecognizer:gestureRecognizer];
        view.__fw_drawerView = self;
    }
    return self;
}

#pragma mark - Accessor

- (void)setDirection:(UISwipeGestureRecognizerDirection)direction
{
    _direction = direction;
    _position = self.isVertical ? self.view.frame.origin.y : self.view.frame.origin.x;
}

- (void)setPositions:(NSArray<NSNumber *> *)positions
{
    if (positions.count < 2) return;
    
    _positions = [positions sortedArrayUsingComparator:^NSComparisonResult(NSNumber *obj1, NSNumber *obj2) {
        return [obj1 compare:obj2];
    }];
}

- (void)setScrollView:(UIScrollView *)scrollView
{
    if (scrollView != _scrollView) {
        if (_scrollView && _scrollView.delegate == self) {
            _scrollView.delegate = nil;
        }
        _scrollView = scrollView;
        if (scrollView && !scrollView.delegate) {
            scrollView.delegate = self;
        }
    }
}

- (BOOL)enabled
{
    return self.gestureRecognizer.enabled;
}

- (void)setEnabled:(BOOL)enabled
{
    self.gestureRecognizer.enabled = enabled;
}

- (CGFloat)openPosition
{
    return self.isReverse ? self.positions.firstObject.doubleValue : self.positions.lastObject.doubleValue;
}

- (CGFloat)middlePosition
{
    return [self positionAtIndex:self.positions.count / 2];
}

- (CGFloat)closePosition
{
    return self.isReverse ? self.positions.lastObject.doubleValue : self.positions.firstObject.doubleValue;
}

#pragma mark - Private

- (BOOL)isVertical
{
    return self.direction == UISwipeGestureRecognizerDirectionUp || self.direction == UISwipeGestureRecognizerDirectionDown;
}

- (BOOL)isReverse
{
    return self.direction == UISwipeGestureRecognizerDirectionUp || self.direction == UISwipeGestureRecognizerDirectionLeft;
}

- (UIRectEdge)scrollEdge
{
    switch (self.direction) {
        case UISwipeGestureRecognizerDirectionUp:
            return UIRectEdgeTop;
        case UISwipeGestureRecognizerDirectionDown:
            return UIRectEdgeBottom;
        case UISwipeGestureRecognizerDirectionLeft:
            return UIRectEdgeLeft;
        case UISwipeGestureRecognizerDirectionRight:
            return UIRectEdgeRight;
        default:
            return UIRectEdgeNone;
    }
}

- (BOOL)isDirection:(UIPanGestureRecognizer *)gestureRecognizer
{
    UISwipeGestureRecognizerDirection direction = gestureRecognizer.__fw_swipeDirection;
    switch (self.direction) {
        case UISwipeGestureRecognizerDirectionUp:
            return direction == UISwipeGestureRecognizerDirectionDown;
        case UISwipeGestureRecognizerDirectionDown:
            return direction == UISwipeGestureRecognizerDirectionUp;
        case UISwipeGestureRecognizerDirectionLeft:
            return direction == UISwipeGestureRecognizerDirectionRight;
        case UISwipeGestureRecognizerDirectionRight:
            return direction == UISwipeGestureRecognizerDirectionLeft;
        default:
            return NO;
    }
}

- (CGFloat)nextPosition
{
    __block CGFloat position;
    if (self.position > self.originPosition) {
        [self.positions enumerateObjectsUsingBlock:^(NSNumber *obj, NSUInteger idx, BOOL *stop) {
            CGFloat maxKickback = (obj.doubleValue == self.positions.lastObject.doubleValue) ? obj.doubleValue : obj.doubleValue + self.kickbackHeight;
            if (self.position <= maxKickback) {
                position = obj.doubleValue;
                *stop = YES;
            }
        }];
    } else {
        [self.positions enumerateObjectsWithOptions:NSEnumerationReverse usingBlock:^(NSNumber *obj, NSUInteger idx, BOOL *stop) {
            CGFloat minKickback = (obj.doubleValue == self.positions.firstObject.doubleValue) ? obj.doubleValue : obj.doubleValue - self.kickbackHeight;
            if (self.position >= minKickback) {
                position = obj.doubleValue;
                *stop = YES;
            }
        }];
    }
    return position;
}

- (BOOL)canScroll:(UIScrollView *)scrollView
{
    if (self.scrollViewFilter) return self.scrollViewFilter(scrollView);
    if (!scrollView.__fw_isViewVisible || !scrollView.scrollEnabled) return NO;
    if (self.isVertical) {
        if (![scrollView __fw_canScrollVertical]) return NO;
    } else {
        if (![scrollView __fw_canScrollHorizontal]) return NO;
    }
    return YES;
}

- (void)togglePosition:(CGFloat)position
{
    self.view.frame = CGRectMake(
        self.isVertical ? self.view.frame.origin.x : position,
        self.isVertical ? position : self.view.frame.origin.y,
        self.view.frame.size.width,
        self.view.frame.size.height
    );
}

- (void)notifyPosition:(BOOL)finished
{
    [self adjustScrollInset];
    
    if (self.positionChanged) {
        self.positionChanged(self.position, finished);
    }
    if ([self.delegate respondsToSelector:@selector(drawerView:positionChanged:finished:)]) {
        [self.delegate drawerView:self positionChanged:self.position finished:finished];
    }
}

- (void)adjustScrollInset
{
    NSArray *insets = self.scrollViewInsets && self.scrollView ? self.scrollViewInsets(self.scrollView) : nil;
    if (insets.count > 0 && insets.count == self.positions.count) {
        [self.positions enumerateObjectsUsingBlock:^(NSNumber * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            NSNumber *next = idx < (self.positions.count - 1) ? self.positions[idx + 1] : nil;
            if ((next && self.position < next.doubleValue) || !next) {
                UIEdgeInsets inset = [insets[idx] UIEdgeInsetsValue];
                if (!UIEdgeInsetsEqualToEdgeInsets(self.scrollView.contentInset, inset)) {
                    self.scrollView.contentInset = inset;
                }
                *stop = YES;
            }
        }];
    }
}

- (void)animateComplete:(CGFloat)position
{
    // 动画完成时需释放displayLink
    if (self.displayLink) {
        [self.displayLink invalidate];
        self.displayLink = nil;
    }
    
    [self togglePosition:position];
    self.position = position;
    [self notifyPosition:YES];
}

#pragma mark - Public

- (void)setPosition:(CGFloat)position animated:(BOOL)animated
{
    if (self.position == position) return;
    
    // 不执行动画
    if (!animated) {
        [self togglePosition:position];
        self.position = position;
        [self notifyPosition:YES];
        return;
    }
    
    // 使用CADisplayLink监听动画过程中的位置
    if (self.displayLink) {
        [self.displayLink invalidate];
        self.displayLink = nil;
    }
    self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(displayLinkAction)];
    [self.displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
    
    // 执行动画移动到指定位置，动画完成标记拖拽位置并回调
    if (self.animationBlock) {
        __weak __typeof__(self) self_weak_ = self;
        self.animationBlock(^{
            __typeof__(self) self = self_weak_;
            [self togglePosition:position];
        }, ^(BOOL finished) {
            __typeof__(self) self = self_weak_;
            [self animateComplete:position];
        });
    } else {
        [UIView animateWithDuration:0.3 delay:0 usingSpringWithDamping:0.75 initialSpringVelocity:0 options:UIViewAnimationOptionBeginFromCurrentState|UIViewAnimationOptionCurveEaseInOut animations:^{
            [self togglePosition:position];
        } completion:^(BOOL finished) {
            [self animateComplete:position];
        }];
    }
}

- (CGFloat)positionAtIndex:(NSInteger)index
{
    if (index < 0 || index >= self.positions.count) return 0;
    return [self.positions[index] doubleValue];
}

- (BOOL)isPositionIndex:(NSInteger)index
{
    if (index < 0 || index >= self.positions.count) return NO;
    return self.position == [self.positions[index] doubleValue];
}

- (void)setPositionIndex:(NSInteger)index animated:(BOOL)animated
{
    if (index < 0 || index >= self.positions.count) return;
    [self setPosition:[self.positions[index] doubleValue] animated:animated];
}

- (void)displayLinkAction
{
    // 监听动画过程中的位置，访问view.layer.presentationLayer即可
    self.position = self.isVertical ? self.view.layer.presentationLayer.frame.origin.y : self.view.layer.presentationLayer.frame.origin.x;
    [self notifyPosition:NO];
}

- (void)gestureRecognizerAction:(UIPanGestureRecognizer *)gestureRecognizer
{
    switch (gestureRecognizer.state) {
        // 拖动开始时记录起始位置信息
        case UIGestureRecognizerStateBegan: {
            self.position = self.isVertical ? self.view.frame.origin.y : self.view.frame.origin.x;
            self.originPosition = self.position;
            
            self.isOriginScrollView = [gestureRecognizer __fw_hitTestWithView:self.scrollView];
            self.isOriginDirection = [self isDirection:gestureRecognizer] || [self isDirection:self.scrollView.panGestureRecognizer];
            self.originOffset = self.scrollView.contentOffset;
            self.isOriginDraggable = self.isOriginDirection && [self.scrollView __fw_isScrollTo:self.scrollEdge];
            NSArray *positions = self.scrollViewPositions && self.scrollView ? self.scrollViewPositions(self.scrollView) : nil;
            self.isOriginScrollable = self.originPosition == self.openPosition || [positions containsObject:@(self.originPosition)];
            break;
        }
        // 拖动改变时更新视图位置
        case UIGestureRecognizerStateChanged: {
            // 记录并清空相对父视图的移动距离
            CGPoint transition = [gestureRecognizer translationInView:self.view.superview];
            [gestureRecognizer setTranslation:CGPointZero inView:self.view.superview];
                
            // 视图跟随拖动移动指定距离，且移动时限制不超过范围
            CGFloat position = self.isVertical ? (self.view.frame.origin.y + transition.y) : (self.view.frame.origin.x + transition.x);
            if (position < self.positions.firstObject.doubleValue) {
                position = self.positions.firstObject.doubleValue;
            } else if (position > self.positions.lastObject.doubleValue) {
                position = self.positions.lastObject.doubleValue;
            }
                
            // 执行位移并回调
            [self togglePosition:position];
            self.position = position;
            [self gestureRecognizerDidScroll];
            [self notifyPosition:NO];
            break;
        }
        // 拖动结束时停留指定位置
        case UIGestureRecognizerStateFailed:
        case UIGestureRecognizerStateEnded: {
            // 停留位置未发生改变时不执行动画，直接回调
            if (self.position == self.originPosition) {
                [self notifyPosition:YES];
            // 停留位置发生改变时执行动画，动画完成后回调
            } else {
                [self setPosition:self.nextPosition animated:YES];
            }
            break;
        }
        default:
            break;
    }
}

#pragma mark - UIScrollViewDelegate

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (scrollView != self.scrollView || !self.gestureRecognizer.__fw_isActive) return;
    if (![self canScroll:self.scrollView]) return;
    
    NSArray *positions = self.scrollViewPositions ? self.scrollViewPositions(self.scrollView) : nil;
    if (positions.count > 0) {
        self.panDisabled = NO;
        if (self.isOriginScrollable) {
            if (self.isOriginDraggable) {
                [self.scrollView __fw_scrollTo:self.scrollEdge animated:NO];
            } else {
                [self togglePosition:self.originPosition];
                self.position = self.originPosition;
            }
        } else {
            self.scrollView.contentOffset = self.originOffset;
        }
        return;
    }
    
    if ([self.scrollView __fw_isScrollTo:self.scrollEdge]) {
        self.panDisabled = NO;
    }
    if (!self.panDisabled) {
        [self.scrollView __fw_scrollTo:self.scrollEdge animated:NO];
    }
}

- (void)gestureRecognizerDidScroll
{
    if (!self.scrollView || !self.gestureRecognizer.enabled) return;
    if (![self canScroll:self.scrollView]) return;
    
    NSArray *positions = self.scrollViewPositions ? self.scrollViewPositions(self.scrollView) : nil;
    if (positions.count > 0) {
        self.panDisabled = NO;
        if (self.isOriginScrollable) {
            if (!self.isOriginDraggable && self.isOriginScrollView) {
                [self togglePosition:self.originPosition];
                self.position = self.originPosition;
            }
        }
        return;
    }
    
    if (self.position == self.openPosition) {
        self.panDisabled = !self.isOriginDraggable && (self.isOriginScrollView || !self.isOriginDirection);
    }
    if (self.panDisabled) {
        [self togglePosition:self.openPosition];
        self.position = self.openPosition;
    } else {
        [self.scrollView __fw_scrollTo:self.scrollEdge animated:NO];
    }
}

#pragma mark - UIGestureRecognizerDelegate

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer
{
    if ([otherGestureRecognizer isKindOfClass:[UIPanGestureRecognizer class]] &&
        [otherGestureRecognizer.view isKindOfClass:[UIScrollView class]]) {
        if (self.autoDetected) {
            UIScrollView *scrollView = (UIScrollView *)otherGestureRecognizer.view;
            if ([self canScroll:scrollView]) {
                self.scrollView = scrollView;
                return YES;
            }
        } else {
            if (self.scrollView && self.scrollView == otherGestureRecognizer.view &&
                [self canScroll:self.scrollView]) {
                return YES;
            }
        }
    }
    return NO;
}

@end

File: Sources/FWFramework/Module/View/DrawerView.swift
//
//  DrawerView.swift
//  FWFramework
//
//  Created by wuyong on 2022/8/23.
//

import UIKit

// MARK: - Wrapper+UIView
/// 视图抽屉拖拽效果分类
@MainActor extension Wrapper where Base: UIView {
    /// 抽屉拖拽视图，绑定抽屉拖拽效果后才存在
    public var drawerView: DrawerView? {
        get {
            property(forName: "drawerView") as? DrawerView
        }
        set {
            setProperty(newValue, forName: "drawerView")
        }
    }

    /**
     设置抽屉拖拽效果。如果view为滚动视图，自动处理与滚动视图pan手势冲突的问题

     @param direction 拖拽方向，如向上拖动视图时为Up，默认向上
     @param positions 抽屉位置，至少两级，相对于view父视图的originY位置
     @param kickbackHeight 回弹高度，拖拽小于该高度执行回弹
     @param positionChanged 抽屉视图位移回调，参数为相对父视图的origin位置和是否拖拽完成的标记
     @return 抽屉拖拽视图
     */
    @discardableResult
    public func drawerView(
        _ direction: UISwipeGestureRecognizer.Direction,
        positions: [CGFloat],
        kickbackHeight: CGFloat,
        positionChanged: ((CGFloat, Bool) -> Void)? = nil
    ) -> DrawerView {
        let drawerView = DrawerView(view: base)
        if direction.rawValue > 0 {
            drawerView.direction = direction
        }
        drawerView.positions = positions
        drawerView.kickbackHeight = kickbackHeight
        drawerView.positionChanged = positionChanged
        return drawerView
    }
}

// MARK: - Wrapper+UIScrollView
/// 滚动视图纵向手势冲突无缝滑动分类，需允许同时识别多个手势
@MainActor extension Wrapper where Base: UIScrollView {
    /// 外部滚动视图是否位于顶部固定位置，在顶部时不能滚动
    public var drawerSuperviewFixed: Bool {
        get { propertyBool(forName: "drawerSuperviewFixed") }
        set { setPropertyBool(newValue, forName: "drawerSuperviewFixed") }
    }

    /// 外部滚动视图scrollViewDidScroll调用，参数为固定的位置
    public func drawerSuperviewDidScroll(_ position: CGFloat) {
        if base.contentOffset.y >= position {
            drawerSuperviewFixed = true
        }
        if drawerSuperviewFixed {
            base.contentOffset = CGPoint(x: base.contentOffset.x, y: position)
        }
    }

    /// 内嵌滚动视图scrollViewDidScroll调用，参数为外部滚动视图
    public func drawerSubviewDidScroll(_ superview: UIScrollView) {
        if base.contentOffset.y <= 0 {
            superview.fw.drawerSuperviewFixed = false
        }
        if !superview.fw.drawerSuperviewFixed {
            base.contentOffset = CGPoint(x: base.contentOffset.x, y: 0)
        }
    }
}

// MARK: - DrawerView
/// 抽屉拖拽视图事件代理
@MainActor public protocol DrawerViewDelegate: AnyObject {
    /// 抽屉视图位移回调，参数为相对view父视图的origin位置和是否拖拽完成的标记
    func drawerView(_ drawerView: DrawerView, positionChanged position: CGFloat, finished: Bool)
}

/// 抽屉拖拽视图
open class DrawerView: NSObject, UIGestureRecognizerDelegate, UIScrollViewDelegate {
    /// 事件代理，默认nil
    open weak var delegate: DrawerViewDelegate?

    /// 拖拽方向，如向上拖动视图时为Up，向下为Down，向右为Right，向左为Left。默认向上
    open var direction: UISwipeGestureRecognizer.Direction = .up {
        didSet {
            if let view {
                position = isVertical ? view.frame.origin.y : view.frame.origin.x
            }
        }
    }

    /// 抽屉位置，至少两级，相对于view父视图的originY位置，自动从小到大排序
    open var positions: [CGFloat] {
        get {
            _positions
        }
        set {
            if newValue.count < 2 { return }
            _positions = newValue.sorted(by: { $0 < $1 })
        }
    }

    private var _positions: [CGFloat] = []

    /// 回弹高度，拖拽小于该高度执行回弹，默认为0
    open var kickbackHeight: CGFloat = 0

    /// 是否启用拖拽，默认YES。其实就是设置手势的enabled
    open var enabled: Bool {
        get { gestureRecognizer.isEnabled }
        set { gestureRecognizer.isEnabled = newValue }
    }

    /// 是否自动检测滚动视图，默认YES。如需手工指定，请禁用之
    open var autoDetected: Bool = true

    /// 指定滚动视图，自动处理与滚动视图pan手势在指定方向的冲突。先尝试设置delegate为自身，尝试失败请手工调用scrollViewDidScroll
    open weak var scrollView: UIScrollView? {
        didSet {
            if let drawerView = oldValue?.delegate as? DrawerView, drawerView == self {
                oldValue?.delegate = nil
            }
            if let scrollView, scrollView.delegate == nil {
                scrollView.delegate = self
            }
        }
    }

    /// 抽屉视图，自动添加pan手势
    open private(set) weak var view: UIView?

    /// 抽屉拖拽手势，默认设置delegate为自身
    open private(set) lazy var gestureRecognizer: UIPanGestureRecognizer = {
        let result = UIPanGestureRecognizer(target: self, action: #selector(gestureRecognizerAction(_:)))
        result.delegate = self
        return result
    }()

    /// 抽屉视图当前位置
    open private(set) var position: CGFloat = 0

    /// 抽屉视图打开位置
    open var openPosition: CGFloat {
        (isReverse ? positions.first : positions.last) ?? .zero
    }

    /// 抽屉视图中间位置，建议单数时调用
    open var middlePosition: CGFloat {
        position(at: positions.count / 2)
    }

    /// 抽屉视图关闭位置
    open var closePosition: CGFloat {
        (isReverse ? positions.last : positions.first) ?? .zero
    }

    /// 抽屉视图位移回调，参数为相对view父视图的origin位置和是否拖拽完成的标记
    open var positionChanged: ((_ position: CGFloat, _ finished: Bool) -> Void)?

    /// 自定义动画句柄，动画必须调用animations和completion句柄
    open var animationBlock: ((_ animations: () -> Void, _ completion: (Bool) -> Void) -> Void)?

    /// 滚动视图过滤器，默认只处理可滚动视图的冲突。如需其它条件，可自定义此句柄
    open var scrollViewFilter: ((UIScrollView) -> Bool)?

    /// 自定义滚动视图允许滚动的位置，默认nil时仅openPosition可滚动
    open var scrollViewPositions: ((UIScrollView) -> [CGFloat])?

    /// 自定义滚动视图在各个位置的contentInset(从小到大，数量同positions)，默认nil时不处理。UITableView时也可使用tableFooterView等实现
    open var scrollViewInsets: ((UIScrollView) -> [UIEdgeInsets])?

    private var displayLink: CADisplayLink?
    private var panDisabled = false
    private var originPosition: CGFloat = .zero
    private var originOffset: CGPoint = .zero
    private var isOriginDraggable = false
    private var isOriginScrollable = false
    private var isOriginScrollView = false
    private var isOriginDirection = false

    private var isVertical: Bool {
        direction == .up || direction == .down
    }

    private var isReverse: Bool {
        direction == .up || direction == .left
    }

    private var scrollEdge: UIRectEdge {
        switch direction {
        case .up:
            return .top
        case .down:
            return .bottom
        case .left:
            return .left
        case .right:
            return .right
        default:
            return []
        }
    }

    // MARK: - Lifecycle
    /// 创建抽屉拖拽视图，view会强引用之。view为滚动视图时，详见scrollView属性
    public init(view: UIView) {
        super.init()
        self.view = view
        self.position = isVertical ? view.frame.origin.y : view.frame.origin.x

        if let scrollView = view as? UIScrollView {
            self.scrollView = scrollView
            if scrollView.delegate == nil {
                scrollView.delegate = self
            }
        }

        view.addGestureRecognizer(gestureRecognizer)
        view.fw.drawerView = self
    }

    // MARK: - Public
    /// 设置抽屉效果视图到指定位置，如果位置发生改变，会触发抽屉callback回调
    open func setPosition(_ position: CGFloat, animated: Bool = true) {
        guard self.position != position else { return }

        // 不执行动画
        if !animated {
            togglePosition(position)
            self.position = position
            notifyPosition(true)
            return
        }

        // 使用CADisplayLink监听动画过程中的位置
        if displayLink != nil {
            displayLink?.invalidate()
            displayLink = nil
        }
        displayLink = CADisplayLink(target: self, selector: #selector(displayLinkAction))
        displayLink?.add(to: .current, forMode: .common)

        // 执行动画移动到指定位置，动画完成标记拖拽位置并回调
        if let animationBlock {
            animationBlock({ [weak self] in
                self?.togglePosition(position)
            }, { [weak self] _ in
                self?.animateComplete(position)
            })
        } else {
            UIView.animate(withDuration: 0.3, delay: 0, usingSpringWithDamping: 0.75, initialSpringVelocity: 0, options: [.beginFromCurrentState, .curveEaseInOut], animations: {
                self.togglePosition(position)
            }, completion: { _ in
                self.animateComplete(position)
            })
        }
    }

    /// 获取抽屉视图指定索引位置(从小到大)，获取失败返回0
    open func position(at index: Int) -> CGFloat {
        guard index >= 0, index < positions.count else { return 0 }
        return positions[index]
    }

    /// 判断当前抽屉效果视图是否在指定索引位置(从小到大)
    open func isPosition(at index: Int) -> Bool {
        guard index >= 0, index < positions.count else { return false }
        return position == positions[index]
    }

    /// 设置抽屉效果视图到指定索引位置(从小到大)，如果位置发生改变，会触发抽屉callback回调
    open func setPosition(at index: Int, animated: Bool = true) {
        guard index >= 0, index < positions.count else { return }
        setPosition(positions[index], animated: animated)
    }

    // MARK: - UIScrollViewDelegate
    /// 如果scrollView已自定义delegate，需在scrollViewDidScroll手工调用本方法
    open func scrollViewDidScroll(_ scrollView: UIScrollView) {
        guard scrollView == self.scrollView, gestureRecognizer.isEnabled else { return }
        guard canScroll(scrollView) else { return }

        let positions = scrollViewPositions?(scrollView)
        if positions?.count ?? 0 > 0 {
            panDisabled = false
            if isOriginScrollable {
                if isOriginDraggable {
                    scrollView.fw.scroll(to: scrollEdge, animated: false)
                } else {
                    togglePosition(originPosition)
                    position = originPosition
                }
            } else {
                scrollView.contentOffset = originOffset
            }
            return
        }

        if scrollView.fw.isScroll(to: scrollEdge) {
            panDisabled = false
        }
        if !panDisabled {
            scrollView.fw.scroll(to: scrollEdge, animated: false)
        }
    }

    // MARK: - UIGestureRecognizerDelegate
    open func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer, shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
        if otherGestureRecognizer is UIPanGestureRecognizer,
           let otherView = otherGestureRecognizer.view as? UIScrollView {
            if autoDetected {
                if canScroll(otherView) {
                    scrollView = otherView
                    return true
                }
            } else {
                if let scrollView, scrollView == otherView, canScroll(scrollView) {
                    return true
                }
            }
        }
        return false
    }

    // MARK: - Private
    private func isDirection(_ gestureRecognizer: UIPanGestureRecognizer) -> Bool {
        let swipeDirection = gestureRecognizer.fw.swipeDirection
        switch direction {
        case .up:
            return swipeDirection == .down
        case .down:
            return swipeDirection == .up
        case .left:
            return swipeDirection == .right
        case .right:
            return swipeDirection == .left
        default:
            return false
        }
    }

    private var nextPosition: CGFloat {
        var nextPosition: CGFloat = .zero
        if position > originPosition {
            for obj in positions {
                let maxKickback = (obj == positions.last) ? obj : obj + kickbackHeight
                if position <= maxKickback {
                    nextPosition = obj
                    break
                }
            }
        } else {
            for obj in positions.reversed() {
                let minKickback = (obj == positions.first) ? obj : obj - kickbackHeight
                if position >= minKickback {
                    nextPosition = obj
                    break
                }
            }
        }
        return nextPosition
    }

    private func canScroll(_ scrollView: UIScrollView) -> Bool {
        if let scrollViewFilter { return scrollViewFilter(scrollView) }
        if !scrollView.fw.isViewVisible || !scrollView.isScrollEnabled { return false }
        if isVertical {
            if !scrollView.fw.canScrollVertical { return false }
        } else {
            if !scrollView.fw.canScrollHorizontal { return false }
        }
        return true
    }

    private func togglePosition(_ position: CGFloat) {
        guard let view else { return }
        view.frame = CGRect(
            x: isVertical ? view.frame.origin.x : position,
            y: isVertical ? position : view.frame.origin.y,
            width: view.frame.size.width,
            height: view.frame.size.height
        )
    }

    private func notifyPosition(_ finished: Bool) {
        adjustScrollInset()

        positionChanged?(position, finished)
        delegate?.drawerView(self, positionChanged: position, finished: finished)
    }

    private func adjustScrollInset() {
        guard let scrollView, let insets = scrollViewInsets?(scrollView) else { return }
        if insets.count > 0 && insets.count == positions.count {
            for (idx, _) in positions.enumerated() {
                let next = idx < (positions.count - 1) ? positions[idx + 1] : nil
                if (next != nil && position < next!) || next == nil {
                    let inset = insets[idx]
                    if scrollView.contentInset != inset {
                        scrollView.contentInset = inset
                    }
                    break
                }
            }
        }
    }

    private func animateComplete(_ position: CGFloat) {
        // 动画完成时需释放displayLink
        if displayLink != nil {
            displayLink?.invalidate()
            displayLink = nil
        }

        togglePosition(position)
        self.position = position
        notifyPosition(true)
    }

    @objc private func displayLinkAction() {
        // 监听动画过程中的位置，访问view.layer.presentation即可
        position = (isVertical ? view?.layer.presentation()?.frame.origin.y : view?.layer.presentation()?.frame.origin.x) ?? .zero
        notifyPosition(false)
    }

    @objc private func gestureRecognizerAction(_ gestureRecognizer: UIPanGestureRecognizer) {
        switch gestureRecognizer.state {
        // 拖动开始时记录起始位置信息
        case .began:
            position = (isVertical ? view?.frame.origin.y : view?.frame.origin.x) ?? .zero
            originPosition = position

            isOriginScrollView = gestureRecognizer.fw.hitTest(view: scrollView)
            isOriginDirection = isDirection(gestureRecognizer) || (scrollView != nil && isDirection(scrollView!.panGestureRecognizer))
            originOffset = scrollView?.contentOffset ?? .zero
            isOriginDraggable = isOriginDirection && (scrollView?.fw.isScroll(to: scrollEdge) ?? false)
            let positions = scrollView != nil ? scrollViewPositions?(scrollView!) : nil
            isOriginScrollable = originPosition == openPosition || positions?.contains(originPosition) == true
        // 拖动改变时更新视图位置
        case .changed:
            // 记录并清空相对父视图的移动距离
            let transition = gestureRecognizer.translation(in: view?.superview)
            gestureRecognizer.setTranslation(.zero, in: view?.superview)

            // 视图跟随拖动移动指定距离，且移动时限制不超过范围
            var position = isVertical ? ((view?.frame.origin.y ?? 0) + transition.y) : ((view?.frame.origin.x ?? 0) + transition.x)
            if position < (positions.first ?? 0) {
                position = (positions.first ?? 0)
            } else if position > (positions.last ?? 0) {
                position = (positions.last ?? 0)
            }

            // 执行位移并回调
            togglePosition(position)
            self.position = position
            gestureRecognizerDidScroll()
            notifyPosition(false)
        // 拖动结束时停留指定位置
        case .failed, .ended:
            // 停留位置未发生改变时不执行动画，直接回调
            if position == originPosition {
                notifyPosition(true)
                // 停留位置发生改变时执行动画，动画完成后回调
            } else {
                setPosition(nextPosition, animated: true)
            }
        default:
            break
        }
    }

    private func gestureRecognizerDidScroll() {
        guard let scrollView, gestureRecognizer.isEnabled else { return }
        guard canScroll(scrollView) else { return }

        let positions = scrollViewPositions?(scrollView)
        if positions?.count ?? 0 > 0 {
            panDisabled = false
            if isOriginScrollable {
                if !isOriginDraggable && isOriginScrollView {
                    togglePosition(originPosition)
                    position = originPosition
                }
            }
            return
        }

        if position == openPosition {
            panDisabled = !isOriginDraggable && (isOriginScrollView || !isOriginDirection)
        }
        if panDisabled {
            togglePosition(openPosition)
            position = openPosition
        } else {
            scrollView.fw.scroll(to: scrollEdge, animated: false)
        }
    }
}
